test/distributed/elastic/agent/server/test/local_elastic_agent_test.py

```diff
@@ -963,6 +1039,56 @@ def test_double_agent_fault_tolerance_etcd_v2(self):
             backend="etcd-v2", test_to_run=self.double_agent_fault_tolerance
         )
 
+    def double_agent_clean_exit(self):
+        """
+        start ``nnodes`` agents, kill and restart odd ones, validate fault-tolerance works
+        """
+        nnodes = 2
+        wait = 20
+        node_conf = Conf(
+            entrypoint=_bipolar_sleep_function, args=(wait,), local_world_size=2
+        )
+        agent_results = mp.Queue()
+        agent_args = {
+            "conf": node_conf,
+            "agent_results": agent_results,
+            "min_nodes": nnodes,
+            "max_nodes": nnodes,
+            "max_restarts": 0,
+            "exit_barrier_timeout": 300,
+        }
+
+        procs = []
+        for _ in range(nnodes):
+            p = mp.Process(
+                target=self.run_agent,
+                kwargs=agent_args,
+            )
+            procs.append(p)
+            p.start()
+
+        # wait for all processes to finish should not take exit barrier time
+        time_between_processes = 0
+        for i in range(nnodes):
+            p = procs[i]
+            p.join()
+            self.assertNotEqual(0, p.exitcode)
+            time_between_processes = time.monotonic() - time_between_processes
+        # Validate that the processes finish close to each other.
+        # Using a slightly higher timeout than 2 * monitor_interval to make it less flaky
+        self.assertGreater(
+            1, time_between_processes, "Agents are not cleaned up until 2 * monitor_interval"
+        )
+
+    @unittest.skipIf(
+        TEST_WITH_DEV_DBG_ASAN or TEST_WITH_TSAN,
+        "test incompatible with dev/dbg asan or tsan",
+    )
+    def test_double_agent_clean_exit_c10d(self):
```

Comment:
the particular rdzv backend is irrelevant to this test so I'd rename this test to `test_no_exit_barrier_on_failure"